feat(layouts): decompose and centralize layout views

hypeJunction opened the pull request:
(replaced by #10333)

**feat(layouts): decompose and centralize layout views**

Centralizes layouts in a default layout providing better control over which layout elements should be rendered. Decomposes default layout view into smaller elements.

BREAKING CHANGE
one_column, one_sidebar, two_sidebar and content layout views have been removed - plugins should start using the default layout. Plugins need to update their use of page/layouts/default and
page/layouts/elements/filter.

**feat(groups): update group listings to new layout API** 

BREAKING CHANGE
groups/groups_sort_menu view has been removed. Plugins that were using it to alter filter tabs on groups/all page, should instead use "register, menu:filter:groups/all" plugin hook.

hypeJunction committed:
feat(layouts): decompose and centralize layout views

Centralizes layouts in a default layout providing better control
over which layout elements should be rendered. Decomposes default
layout view into smaller elements.

BREAKING CHANGE
one_column, one_sidebar, two_sidebar and content layout views have been
removed - plugins should start using the default layout.
Plugins need to update their use of page/layouts/default and
page/layouts/elements/filter.

hypeJunction committed:
feat(groups): update group listings to new layout API

BREAKING CHANGE
groups/groups_sort_menu view has been removed. Plugins that were using it
to alter filter tabs on groups/all page, should instead use
"register, menu:filter:groups/all" plugin hook.

hypeJunction commented:
So, here is a radical idea! 💩

hypeJunction commented on docs/guides/upgrading.rst:
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
+This takes away control over visual presentation from plugins and hands that control over to the themes. For maximum backwards compatibility, calls to ``elgg_view_layout()`` with these layout names will still yield expected output, but the plugins should start using the ``default`` layout.
+Themes overriding any of the above layout views have to be updated to properly render ALL layout elements.
+
+Themes and plugins overriding ``layouts/elements/filter`` view has to update their usage of view parameters.
```
`page/layouts/elements/filter`
have to

hypeJunction commented on views/default/page/layouts/error.php:
```diff
@@ -7,11 +7,8 @@
+
 $vars['nav'] = ''; // prevents outputting breadcrumbs
 
 echo elgg_view('page/layouts/one_column', $vars);
```
Update to use `default` layout

hypeJunction replied:
No sidebar

hypeJunction commented on views/default/page/layouts/default.php:
```diff
@@ -1,36 +1,58 @@
-echo '</div>';
+echo elgg_format_element('div', [
+	'class' => $class,
+], $sidebar . $sidebar_alt . $body);
```
In Aalborg, it's `$sidebar_alt . $body . $sidebar`, but if we are switching to flexbox, we can just use `order`

mrclay commented on docs/guides/upgrading.rst:
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
+
+As a result of these changes, all layouts can now be easily extended to have multiple tabs. Plugins can pass ``filter_id`` parameter that will allow other plugins to hook into ``register, menu:filter:<filter_id>`` hook and add new tabs. If no ``filter_id`` is provided, default ``register, menu:filter`` hook can be used.
+
+Layout views and their underlying views now receive ``identifier`` and ``segments`` of the page being rendered, so they can make decisions about what elements should be rendered.
```
So we're asking these views to do their own routing?

mrclay replied:
...Allowing them. That's fine.

mrclay commented on engine/lib/views.php:
```diff
@@ -693,31 +693,88 @@ function elgg_view_layout($layout_name, $vars = array()) {
 	}
 	$timer->begin([__FUNCTION__]);
 
-	$params = array();
```
Updating docs of this function might help get this idea across

mrclay commented on views/default/page/layouts/elements/sidebar.php:
```diff
@@ -0,0 +1,16 @@
+?>
+<div class="elgg-sidebar elgg-layout-sidebar clearfix">
+	<?= elgg_view('page/elements/sidebar', $vars) ?>
+</div>
```
Curious, why not just add the wrapper to page/elements/sidebar? A consistency thing?

hypeJunction replied:
This seemed cleaner

mrclay commented on docs/guides/upgrading.rst:
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
+
+Themes and plugins overriding ``layouts/elements/filter`` view has to update their usage of view parameters.
+
+As a result of these changes, all layouts can now be easily extended to have multiple tabs. Plugins can pass ``filter_id`` parameter that will allow other plugins to hook into ``register, menu:filter:<filter_id>`` hook and add new tabs. If no ``filter_id`` is provided, default ``register, menu:filter`` hook can be used.
```
👍

mrclay commented:
If I follow, if I supply content to `sidebar_alt` I get the 2nd column automatically. I wonder why we didn't do it that way before.

jdalsem commented:
Finally... one layout to rule them all...

hypeJunction commented:
- Updated the docs
- Threw before and after layout hooks out of the window
- Updated admin/error layouts for consistency
- Updated theme sandbox with examples

mrclay commented on mod/developers/views/default/theme_sandbox/layouts/two_sidebar.php:
```diff
@@ -1,9 +1,24 @@
+		],
+	],
+	'filter_value' => get_input('filter', 'tab1'),
+));
```
Awesome.

mrclay commented:
> Threw before and after layout hooks out of the window

Fair enough.

hypeJunction commented on views/default/page/layouts/default.php:
```diff
@@ -1,36 +1,71 @@
+ * @uses $vars['class']        Additional CSS classes to apply to the layout
+ *
+ * @uses $vars['breadcrumbs']  An array of breadcrumbs (see navigation/breadcrumbs)
+ * @uses $vars['nav']          Alternative page navigation (replaces navigation/breadcrumbs)
```
Should we throw this out? If breadcrumbs is string use it, if array - render navigation/breadcrumbs?

mrclay replied:
SGTM

hypeJunction committed:
chore(navigation): update the use of breadcrumbs in layout views

Removes nav parameter from layout views and improves the use of
breadcrumbs parameter.
Adds current route identifier and segments to the prepare, breadcrumbs
hook params.
Allows injecting breadcrumbs into elgg_get_breadcrumbs()

mrclay commented:
Do you feel this is approaching "finished"? Should I convert one plugin to use this to get the feel of it?

hypeJunction commented:
I feel it's done. There typos in docs that need fixing but I don't see them anymore. Please do adapt something.  River could use some love.

hypeJunction commented:
I have some more plans for layouts/pages, so the sooner we merge this, the better.

hypeJunction commented:
@mrclay Could you make this one a priority, so we can merge it sooner?

Once this is in, I want to try the following:

- Update `elgg_view_page()` to accept an array of layout parameters and teach it to draw layouts:

```php
$params = [
     // Shell params
    `#layout` => 'default',
    '#title' => 'Page title',
    '#shell' => 'default',
    // Layout params + shell vars
    'title' => 'Layout title',
    'content' => 'Layout content',
    'entity' => $entity,
    'page_owner' => $page_owner,
];

echo elgg_view_page($params);
// equals to
echo elgg_view_resource('default', $params);
```

mrclay commented on engine/lib/navigation.php:
```diff
@@ -272,31 +272,50 @@ function elgg_pop_breadcrumb() {
+	array_shift($params['segments']);
+
 	$breadcrumbs = elgg_trigger_plugin_hook('prepare', 'breadcrumbs', $params, $breadcrumbs);
+	_elgg_services()->logger->error('"prepare, breadcrumbs" hook must return an array of breadcrumbs');
```
Move this inside the condition

mrclay reviewed (approved):
Besides moving the error line, I'm ready to go for it.

mrclay commented:
Replaced by #10333
